tests/integration/test_upgrades.py

```diff
@@ -0,0 +1,104 @@
+# Copyright 2023 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""Ranger charm upgrades integration tests."""
+
+import asyncio
+import logging
+
+import pytest
+import pytest_asyncio
+import requests
+import yaml
+from helpers import APP_NAME, METADATA, POSTGRES_NAME, SECURE_PWD, get_unit_url
+from pytest_operator.plugin import OpsTest
+
+logger = logging.getLogger(__name__)
+
+
+@pytest.mark.skip_if_deployed
+@pytest_asyncio.fixture(name="deploy", scope="module")
+async def deploy(ops_test: OpsTest):
+    """Deploy the app."""
+    asyncio.gather(
+        ops_test.model.deploy(POSTGRES_NAME, channel="14", trust=True),
+    )
+
+    ranger_config = {"ranger-admin-password": SECURE_PWD}
+    await ops_test.model.deploy(
+        APP_NAME, channel="stable", config=ranger_config
+    )
+
+    async with ops_test.fast_forward():
+        await ops_test.model.wait_for_idle(
+            apps=[POSTGRES_NAME],
+            status="active",
+            raise_on_blocked=False,
+            timeout=2000,
```

Comment:
Timeout's a bit excessive, if your model times out for whatever reason, this will make the integration test run for longer than necessary.